YAML/YAML.sublime-syntax

```diff
@@ -490,74 +519,94 @@ contexts:
         - include: flow-pair
+        - include: flow-node  # for sets
 
   flow-pair:
     - match: \?
-      scope: punctuation.definition.key-value.begin.yaml
+      scope: meta.mapping.key.yaml punctuation.definition.key.begin.yaml
       push:
-        - meta_scope: meta.flow-pair.explicit.yaml
-        - match: (?=[},\]]) # Empty mapping keys & values are allowed
-          pop: true
-        - include: flow-pair
-        - include: flow-node
-        - match: :(?=\s|$|{{c_flow_indicator}})
-          scope: punctuation.separator.key-value.mapping.yaml
-          set: flow-pair-value
-    # Attempt to match plain-in scalars and highlight as "entity.name.tag",
-    # if followed by a colon
-    - match: |
-        (?x)
-        (?=
-          {{ns_plain_first_plain_in}}
-          (
-              [^\s:{{c_flow_indicator}}]
-            | : [^\s{{c_flow_indicator}}]
-            | \s+ (?![#\s])
-          )*
-          \s*
-          :
-          (\s|$)
-        )
+        - flow-pair-value-set
+        - flow-pair-key
+    # Attempt to match plain-in scalars followed by a colon
+    - match: '{{_flow_key_in_lookahead}}'
       push:
-        # TODO Use a merge type here and add "pop: true" and "scope: entity.name.tag.yaml";
-        # https://github.com/SublimeTextIssues/Core/issues/966
-        - meta_scope: meta.flow-pair.key.yaml
-        - include: flow-scalar-plain-in-implicit-type
-        - match: '{{_flow_scalar_end_plain_in}}'
-          pop: true
-        - match: (?={{ns_plain_first_plain_in}})
-          set:
-            - meta_scope: meta.flow-pair.key.yaml string.unquoted.plain.in.yaml entity.name.tag.yaml
-            - meta_include_prototype: false
-            - match: '{{_flow_scalar_end_plain_in}}'
-              pop: true
-    - include: flow-node
+        - flow-pair-value-set
+        - flow-pair-key
     - match: :(?=\s|$|{{c_flow_indicator}}) # Empty mapping keys allowed
-      scope: meta.flow-pair.yaml punctuation.separator.key-value.mapping.yaml
+      scope: meta.mapping.yaml punctuation.separator.key-value.mapping.yaml
       push: flow-pair-value
 
+  # Variant without clearing the parent scope for pairs in a sequence
+  flow-pair-no-clear:
+    - match: \?
+      scope: meta.mapping.key.yaml punctuation.definition.key.begin.yaml
+      push:
+        - flow-pair-value-set-no-clear
+        - flow-pair-key-no-clear
+    - match: '{{_flow_key_in_lookahead}}'
+      push:
+        - flow-pair-value-set-no-clear
+        - flow-pair-key-no-clear
+    - match: :(?=\s|$|{{c_flow_indicator}})
+      scope: meta.mapping.yaml punctuation.separator.key-value.mapping.yaml
+      push: flow-pair-value-no-clear
+
+  flow-pair-key:
+    - clear_scopes: 1
+    - meta_content_scope: meta.mapping.key.yaml
+    - match: (?=[},\]])  # Empty mapping keys & values are allowed
+      pop: true
+    - match: :(?=\s|$|{{c_flow_indicator}})
+      scope: meta.mapping.yaml punctuation.separator.key-value.mapping.yaml
+      pop: true
```

Comment:
If we don't intend to re-use `flow-pair-key` anywhere else, the `pop: true` could be replaced by `set: flow-pair-value` and the multi-push could be turned into a simple `push: flow-pair-key` in `flow-pair`. The intermediate `flow-pair-value-set` context was obsolete then.

On the other hand, the current solution clearly spearates key and value related contexts with maximum level of reusability, which might be to prefer in such a basic syntax. The proposed change wouldn't result any performance improvements, so I am very fine with the current solution.

Reply:
Handled via https://github.com/FichteForks/Packages/pull/10.